### ISSUE TITLE
Attempting to add a `git` dependency via `flutter pub add` fails

### ISSUE THREAD

**Issue by domesticmouse** [flutter/flutter#144898]
### Steps to reproduce

1. Attempt to add `syntax_highlight` as a dependency, but plot twist, a github hosted fork of the published version
2. Read the documentation from `flutter pub add --help`
3. Attempt to follow the path, and flail hard

### Expected results

Added depency to `pubspec.yaml` that is corrected configured.

### Actual results

Error messages. That I'm not understanding, given the context of my goals.

### Code sample

N/A

### Logs

```console
$ flutter pub add 'syntax_highlight:{"git":"https://github.com/domesticmouse/syntax_highlight","ref":"main"}'
  Using a naked argument for directory is deprecated and will stop working in a future Flutter release.

  Use --directory instead.
Expected to find project root in syntax_highlight:{"git":"https://github.com/domesticmouse/syntax_highlight","ref":"main"}.
```

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.19.3, on macOS 14.4 23E214 darwin-arm64, locale en)
    • Flutter version 3.19.3 on channel stable at /Users/brettmorgan/flutter
    • Upstream repository https://github.com/flutter/flutter
    • Framework revision ba39319843 (3 days ago), 2024-03-07 15:22:21 -0600
    • Engine revision 2e4ba9c6fb
    • Dart version 3.3.1
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/brettmorgan/Library/Android/sdk
    • Platform android-34-ext8, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Community Edition (version 2023.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 77.0.1
    • Dart plugin version 233.13135.65

[✓] VS Code (version 1.87.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 14.4 23E214 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 122.0.6261.112

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by danagbemava-nc** [comment]
Reproducible using the steps outlined above.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.19.3, on macOS 14.3.1 23D60 darwin-arm64, locale en-GB)
    • Flutter version 3.19.3 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ba39319843 (4 days ago), 2024-03-07 15:22:21 -0600
    • Engine revision 2e4ba9c6fb
    • Dart version 3.3.1
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode-15.2.0.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.87.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (5 available)
    • Pixel 7 (mobile)     • adb-28291FDH2001SA-5Lv71w._adb-tls-connect._tcp. • android-arm64  • Android 14 (API 34)
    • Nexus (mobile)       • 00008020-001875E83A38002E                        • ios            • iOS 17.3.1 21D61
    • Dean’s iPad (mobile) • 00008103-000825C811E3401E                        • ios            • iOS 17.3.1 21D61
    • macOS (desktop)      • macos                                            • darwin-arm64   • macOS 14.3.1 23D60 darwin-arm64
    • Chrome (web)         • chrome                                           • web-javascript • Google Chrome 122.0.6261.112

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```
[!] Flutter (Channel master, 3.21.0-2.0.pre.72, on macOS 14.3.1 23D60 darwin-arm64, locale en-GB)
    • Flutter version 3.21.0-2.0.pre.72 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 087b15ae0d (8 hours ago), 2024-03-10 22:33:27 -0400
    • Engine revision d13999c0ca
    • Dart version 3.4.0 (build 3.4.0-220.0.dev)
    • DevTools version 2.33.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode-15.2.0.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.87.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (6 available)
    • Pixel 7 (mobile)                • adb-28291FDH2001SA-5Lv71w._adb-tls-connect._tcp. • android-arm64  • Android 14 (API 34)
    • Nexus (mobile)                  • 00008020-001875E83A38002E                        • ios            • iOS 17.3.1 21D61
    • Dean’s iPad (mobile)            • 00008103-000825C811E3401E                        • ios            • iOS 17.3.1 21D61
    • macOS (desktop)                 • macos                                            • darwin-arm64   • macOS 14.3.1 23D60 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad                            • darwin         • macOS 14.3.1 23D60 darwin-arm64
    • Chrome (web)                    • chrome                                           • web-javascript • Google Chrome 122.0.6261.112

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by christopherfujino** [comment]
@domesticmouse I'm not familiar with that syntax for a `pub add` invocation. Is that supported upstream in `dart pub add`? In other words, is the bug in pub, or in the flutter tool?

**Comment by domesticmouse** [comment]
```console
$ flutter pub add --help
Add dependencies to `pubspec.yaml`.

Invoking `dart pub add foo bar` will add `foo` and `bar` to `pubspec.yaml`
with a default constraint derived from latest compatible version.

Add to dev_dependencies by prefixing with "dev:".

Make dependency overrides by prefixing with "override:".

Add packages with specific constraints or other sources by giving a descriptor
after a colon.

For example:
  * Add a hosted dependency at newest compatible stable version:
    `flutter pub add foo`
  * Add a hosted dev dependency at newest compatible stable version:
    `flutter pub add dev:foo`
  * Add a hosted dependency with the given constraint
    `flutter pub add foo:^1.2.3`
  * Add multiple dependencies:
    `flutter pub add foo dev:bar`
  * Add a path dependency:
    `flutter pub add 'foo:{"path":"../foo"}'`
  * Add a hosted dependency:
    `flutter pub add 'foo:{"hosted":"my-pub.dev"}'`
  * Add an sdk dependency:
    `flutter pub add 'foo:{"sdk":"flutter"}'`
  * Add a git dependency:
    `flutter pub add 'foo:{"git":"https://github.com/foo/foo"}'`
  * Add a dependency override:
    `flutter pub add 'override:foo:1.0.0'`
  * Add a git dependency with a path and ref specified:
    `flutter pub add \
      'foo:{"git":{"url":"../foo.git","ref":"<branch>","path":"<subdir>"}}'`

Usage: dart pub add [options] [<section>:]<package>[:descriptor]
       [<section>:]<package2>[:descriptor] ...]
-h, --help               Print this usage information.
    --[no-]offline       Use cached packages instead of accessing the network.
-n, --dry-run            Report what dependencies would change but don't change
                         any.
    --[no-]precompile    Build executables in immediate dependencies.
-C, --directory=<dir>    Run this in the directory <dir>.

Run "dart help" to see global options.
See https://dart.dev/tools/pub/cmd/pub-add for detailed documentation.
```

**Comment by domesticmouse** [comment]
Attempting to use `dart pub add` instead results in the following:

```console
$ dart pub add 'syntax_highlight:{"git":{"url":"https://github.com/domesticmouse/syntax_highlight.git","ref":"main"}}'

Because super_text_layout_testing depends on syntax_highlight from git which doesn't exist (Could not find git ref 'main' (fatal: cannot use bare repository
  '/Users/brettmorgan/.pub-cache/git/cache/syntax_highlight-9234ce992999c9414c72b26608e4a256e7f9de16' (safe.bareRepository is 'explicit'))), version solving failed.
```

The `flutter pub add` errors appear related.

```console
$ flutter pub add 'syntax_highlight:{"git":{"url":"https://github.com/domesticmouse/syntax_highlight.git","ref":"main"}}'

  Using a naked argument for directory is deprecated and will stop working in a future Flutter release.

  Use --directory instead.
Expected to find project root in syntax_highlight:{"git":{"url":"https://github.com/domesticmouse/syntax_highlight.git","ref":"main"}}.
```

Which I guess means that https://github.com/domesticmouse/syntax_highlight is not correctly formatted for `dart pub add`. I have no idea what I'm doing wrong. Should we bump this issue to the `dart-lang` crew?

**Comment by christopherfujino** [comment]
> Which I guess means that https://github.com/domesticmouse/syntax_highlight is not correctly formatted for `dart pub add`. I have no idea what I'm doing wrong. Should we bump this issue to the `dart-lang` crew?

@jonasfj could you help us out on whether there is a bug in package:syntax_highlight, flutter_tools, pub, or some combination?

**Comment by jonasfj** [comment]
I don't think there is any requirements on repository layout... other than there being a folder that contains the package (including its `pubspec.yaml`).

This seems like a bug in `pub`. Or a wrong input for `dart pub add`

**Comment by jonasfj** [comment]
Correction, I think there is two issues here:

## (A) `dart pub add` doesn't appear to work for `git` in this case

```
$ cd /tmp
$ dart create tt
$ dart pub add 'syntax_highlight:{"git":{"url":"https://github.com/domesticmouse/syntax_highlight.git"}}'
Because tt depends on syntax_highlight from git which doesn't exist (Could not find git ref 'HEAD' (fatal: cannot use bare repository
  '/jonasfj/.pub-cache/git/cache/syntax_highlight-9234ce992999c9414c72b26608e4a256e7f9de16' (safe.bareRepository is 'explicit'))), version solving failed.
```

## (B) flutter_tools legacy support for `<directory>` argument

@sigurdm pointed out that: 
https://github.com/flutter/flutter/blob/c53a18f4e49aa28726bb6e0ea0b525db0c37c750/packages/flutter_tools/lib/src/commands/packages.dart#L261

Basically, it used to be possible to do:
```
flutter pub <cmd> <directory>
```
which was made it impossible to support the completely CLI interface offered by `dart pub`.

As a hack to migrate away from it, `flutter_tools` tests if the first argument contains `/` if so it assumes it might be a directory.

It'd probably be better to migrate away from this.

**Comment by sigurdm** [comment]
I think this is a flutter issue. (try `FLUTTER_ROOT/bin/dart pub add 'syntax_highlight:{"git":{"url":"https://g
ithub.com/domesticmouse/syntax_highlight","ref":"main"}}'`

You can do `flutter pub get <dir>` to operate in a directory. That is something flutter introduced on top of pub.

We wanted to move away from that, and inserted the warning "sing a naked argument for directory is deprecated and will stop working in a future Flutter release." depending upon some heuristic conditions:

https://github.com/flutter/flutter/blob/c53a18f4e49aa28726bb6e0ea0b525db0c37c750/packages/flutter_tools/lib/src/commands/packages.dart#L261

I think it is about time to remove that check and feature.

**Comment by jonasfj** [comment]
Context regarding (B):

I think `pub <cmd>` have always accepted arguments. At-least `pub upgrade <package>` have always been a thing.
But it wasn't very well known, so `flutter_tools` probably without considering the full set of implications decided to make `flutter pub <cmd> <directory>` a thing.

That ofcourse meant that features such as:
 * `pub upgrade <package>`
 * `pub upgrade --major-versions`
 * `pub upgrade --major-versions <package>`
 * `pub add ...`

never worked on `flutter pub`.

So we added a `-C <directory>` flag in `pub`, and tried to detect in `flutter_tools` whether people were calling with `flutter pub <cmd> <directory>`, or if they where trying to pass flags and arguments.

This ofcourse won't work when arguments include `/` or `\`.

It's possible a better choice would be for `flutter_tools` to completely break support for `flutter pub <cmd> <directory>`.

**Comment by sigurdm** [comment]
A) (safe.bareRepository is 'explicit') has been fixed on HEAD: https://github.com/dart-lang/pub/pull/4142

**Comment by AlexV525** [comment]
I've met the same issue when trying to add a dependency with a path config. So I dived into the implementation and found the cause is:
https://github.com/flutter/flutter/blob/97426566104215127a44666fe2b75c18208d1701/packages/flutter_tools/lib/src/commands/packages.dart#L262-L266

Since the argument contains `/`, it will be recognized as an argument of the directory, and then fall into the finding project flow. The solution I proposed in #160223 is to handle the `add` command explicitly with a regular expression check.
